### PR TITLE
Model serialization prototype

### PR DESCRIPTION
**Pull request by rlnx**
1. Serialization engine, see `dal/detail/serialization.hpp`, `dal/backend/serialization.hpp`
2. Data structures serialization: `array`, `table`, `homogen_table`.
3. SVM model serialization
4. Tests

### CODE REVIEW

File: cpp/oneapi/dal/backend/interop/archive.hpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        ONEDAL_ASSERT(!"Not implemented");
+        return daal::services::SharedPtr<daal::byte>{};
+    }
+
+    daal::byte* getArchiveAsArray() override {
+        ONEDAL_ASSERT(!"Not implemented");
+        return nullptr;
+    }
+
+    virtual std::string getArchiveAsString() override {
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/interop/archive.hpp:34]:
```suggestion
    std::string getArchiveAsString() override {
```

File: cpp/oneapi/dal/algo/svm/test/serialization.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            -0.2933, 0.5726,  -0.0968, //
+            0.0617,  1.0032,  -0.4763, //
+            -0.6552, -1.4016, 2.3133, //
+            -0.2874, -0.9861, -0.2129, //
+            0.1375,  -0.326,  -1.3081, //
+            -1.8112, 0.2395,  -0.2247, //
+            1.7241,  -0.4099, -1.7277 //
+        };
+
+        static const std::int32_t y_train_two_cls[] = { 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1,
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/svm/test/serialization.cpp:126]: Why int32, not float type?

File: cpp/oneapi/dal/algo/svm/test/serialization.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        ONEDAL_ASSERT(class_count_ == 2 || class_count_ == 3);
+        const std::int32_t* y_train = (class_count_ == 2) ? y_train_two_cls : y_train_three_cls;
+
+        return { homogen_table::wrap(x_train, row_count, feature_count),
+                 homogen_table::wrap(y_train, row_count, 1) };
+    }
+
+    template <typename T = task_t, detail::enable_if_regression_t<T>* = nullptr>
+    std::tuple<table, table> get_train_data() {
+        // TODO: Replace by regression dataset generator
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/svm/test/serialization.cpp:98]: Are we planning to make such generators?

File: cpp/oneapi/dal/algo/svm/test/serialization.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+private:
+    std::int64_t class_count_ = -1;
+};
+
+using svm_cls_types = COMBINE_TYPES((float, double),
+                                    (svm::method::thunder, svm::method::smo),
+                                    (svm::task::classification));
+
+using svm_reg_types = COMBINE_TYPES((float, double),
+                                    (svm::method::thunder, svm::method::smo),
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/svm/test/serialization.cpp:209]: for regression not support SMO method
```suggestion
                                    (svm::method::thunder),
```

File: cpp/oneapi/dal/test/engine/catch.hpp [hunk 5]
```diff
@@ -29,6 +29,6 @@
 
 // Disables unexpected exceptions handing in Catch2.
 // It is easier to debug exception via GDB if there is handler.
-#define CATCH_CONFIG_DISABLE_EXCEPTIONS
+// #define CATCH_CONFIG_DISABLE_EXCEPTIONS
```
Comment (PetrovKP) [inline cpp/oneapi/dal/test/engine/catch.hpp:32]: don't remember uncomment

File: cpp/oneapi/dal/algo/svm/backend/model_impl.hpp [hunk 6]
```diff
@@ -18,17 +18,20 @@
...
 #include "oneapi/dal/algo/svm/backend/model_interop.hpp"
+#include "oneapi/dal/backend/serialization.hpp"
 
 namespace oneapi::dal::svm {
 
+#define SVM_SERIALIZABLE(Task, ClassificationId, RegressionId)         \
+    ONEDAL_SERIALIZABLE_MAP2(Task,                                     \
+                             (task::classification, ClassificationId), \
+                             (task::regression, RegressionId))
+
```
[the remark this answers is not in the document]
Comment (rlnx) [reply cpp/oneapi/dal/algo/svm/backend/model_impl.hpp:29]: Do you have concern regarding formatting?

File: cpp/oneapi/dal/detail/serialization.hpp [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+}
+
+class binary_output_archive : public base {
+public:
+    void prologue() {}
+    void epilogue() {}
+    void operator()(const void* data, data_type dtype, std::int64_t count = 1) {}
+};
+
+class binary_input_archive : public base {
```
Comment (PetrovKP) [inline cpp/oneapi/dal/detail/serialization.hpp:460]: What are these entities responsible for?

File: cpp/oneapi/dal/test/engine/serialization.hpp [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+    std::int64_t get_count() const {
+        return std::int64_t(get().size());
+    }
+
+private:
+    std::shared_ptr<entries_t> entries_;
+};
+
+class mock_archive_state_reader {
```
[the remark this answers is not in the document]
Comment (rlnx) [reply cpp/oneapi/dal/test/engine/serialization.hpp:81]: I'll use it for testing

File: cpp/oneapi/dal/detail/serialization.hpp (hunk 7 above)
Comment (rlnx) [reply cpp/oneapi/dal/detail/serialization.hpp:490]: Now it's implemented

**Review by PetrovKP — APPROVED** [review]
thanks you!

File: cpp/oneapi/dal/algo/svm/backend/model_impl.hpp [hunk 9]
```diff
@@ -37,10 +40,39 @@ class detail::v1::model_impl : public base {
...
+        ar(support_vectors, coeffs, bias, biases);
+
+        if constexpr (std::is_same_v<Task, task::classification>) {
+            ar(first_class_label, second_class_label, class_count);
+        }
+
+        dal::detail::serialize_polymorphic(interop_, ar);
+    }
+
+    void deserialize(dal::detail::input_archive& ar) override {
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/algo/svm/backend/model_impl.hpp:66]: Just checking: do we allow to deserialize to non-empty model object? Will it flush the previous state completely and replaced by the state from the archive?

File: cpp/oneapi/dal/algo/svm/backend/model_interop.hpp [hunk 10]
```diff
@@ -29,14 +32,27 @@ class model_interop : public base {
...
+
     model_interop_impl(DaalModel& model) : daal_model_(model) {}
 
     const DaalModel get_model() const {
         return daal_model_;
     }
 
+    void serialize(dal::detail::output_archive& ar) const override {
+        dal::backend::interop::daal_output_data_archive daal_ar(ar);
+        daal_ar.setSharedPtrObj(const_cast<DaalModel&>(daal_model_));
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/algo/svm/backend/model_interop.hpp:48]: Seems like `daal_model_` is not a `DaalModel` but a `SharedPtr<DaalModel>` or `DaalModelPtr` at least

File: cpp/oneapi/dal/array.hpp [hunk 11]
```diff
@@ -697,6 +695,15 @@ class array {
...
     }
 
+    void serialize(detail::output_archive& ar) const {
+        impl_->serialize(ar);
+    }
+
+    void deserialize(detail::input_archive& ar) {
+        impl_->deserialize(ar);
+        update_data(impl_.get());
+    }
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/array.hpp:705]: A common question about serialize/deserialize: Imagine one of these operations end up with exception (for example, bad_alloc). Do we have any commitments about the object and archive state after such operation resulted with exception?

File: cpp/oneapi/dal/algo/svm/backend/model_impl.hpp (hunk 9 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/svm/backend/model_impl.hpp:66]: Yes

File: cpp/oneapi/dal/algo/svm/backend/model_interop.hpp (hunk 10 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/svm/backend/model_interop.hpp:48]: `DaalModel` means shared pointer to model, probably need to rename

File: cpp/oneapi/dal/array.hpp (hunk 11 above)
Comment (rlnx) [reply cpp/oneapi/dal/array.hpp:705]: If an exception is raised in the middle of deserialization, there is guaranties regarding valid deserialized object state, such object should not be used.

**Pull request merged**